### PR TITLE
cut and run

### PR DESCRIPTION
**Pull request by lldelisle**
First tentative for CUT&RUN workflow

### CODE REVIEW

**Comment by lldelisle** [comment]
Failed when running bowtie2 with:
`set -o | grep -q pipefail && set -o pipefail;   ln -s '/tmp/tmpc89fczmz/files/000/dataset_3.dat' input_f.fastq &&  ln -s '/tmp/tmpc89fczmz/files/000/dataset_4.dat' input_r.fastq &&    bowtie2  -p ${GALAXY_SLOTS:-4}  -x '/cvmfs/data.galaxyproject.org/managed/bowtie2_index/apiMel4/apiMel4'   -1 'input_f.fastq' -2 'input_r.fastq' -I 0 -X 1000 --fr   --dovetail                --very-sensitive   2> '/tmp/tmpc89fczmz/job_working_directory/000/4/outputs/galaxy_dataset_2928e89f-493a-4171-8cd3-5ae4abd36343.dat'  | samtools sort --no-PG -@${GALAXY_SLOTS:-2} -T "${TMPDIR:-.}" -O bam -o '/tmp/tmpc89fczmz/job_working_directory/000/4/outputs/galaxy_dataset_16ea3644-63a6-4e36-a4cd-7a2c9306845a.dat']`

I don't understand why it is using apiMel4 
In the workflow it is written:
`\"reference_genome\": {\"source\": \"indexed\", \"__current_case__\": 0, \"index\": \"hg38canon\"}`

**Comment by lldelisle** [comment]
TODO before merging:
- set the genome as a input parameter
Check a last time before merging that:
- all parameters are correct.
- there is no 'old' description or whatever that is still here.

**Comment by bgruening** [comment]
Wuhu!!!!!

**Comment by lldelisle** [comment]
Any idea of what is happening in the CI?
image

**Comment by mvdbeek** [comment]
This should mean that one of the uploads failed ... which is strange considering that the rest of the tests succeeded.

**Comment by lldelisle** [comment]
So I just rerun the CI?

**Comment by mvdbeek** [comment]
Worth a try, but I also see
```
"bedGraph error line 299697 of stdin: chromosome chrM has size 16569 but item ends at 16571\nbedGraph error line 299698 of stdin: chromosome chrM has size 16569 but item ends at 16574\nbedGraph error line 299699 of stdin: chromosome chrM has size 16569 but item ends at 16575\nbedGraph error line 299700 of stdin: chromosome chrM has size 16569 but item ends at 16578\nbedGraph error line 299701 of stdin: chromosome chrM has size 16569 but item ends at 16579\nbedGraph error line 299702 of stdin: chromosome chrM has size 16569 but item ends at 16580\nbedGraph error line 299703 of stdin: chromosome chrM has size 16569 but item ends at 16581\nbedGraph error line 299704 of stdin: chromosome chrM has size 16569 but item ends at 16582\nbedGraph error line 299705 of stdin: chromosome chrM has size 16569 but item ends at 16583\nbedGraph error line 299706 of stdin: chromosome chrM has size 16569 but item ends at 16584\nbedGraph error line 299707 of stdin: chromosome chrM has size 16569 but item ends at 16585\nbedGraph error line 299708 of stdin: chromosome chrM has size 16569 but item ends at 16586\nbedGraph error line 299709 of stdin: chromosome chrM has size 16569 but item ends at 16587\nbedGraph error line 299710 of stdin: chromosome chrM has size 16569 but item ends at 16588\nbedGraph error line 299711 of stdin: chromosome chrM has size 16569 but item ends at 16589\nbedGraph error line 299712 of stdin: chromosome chrM has size 16569 but item ends at 16590\nbedGraph error line 299713 of stdin: chromosome chrM has size 16569 but item ends at 16591\nbedGraph error line 299714 of stdin: chromosome chrM has size 16569 but item ends at 16592\nbedGraph error line 299715 of stdin: chromosome chrM has size 16569 but item ends at 16593\nbedGraph error line 299716 of stdin: chromosome chrM has size 16569 but item ends at 16595\nbedGraph error line 299717 of stdin: chromosome chrM has size 16569 but item ends at 16599\nbedGraph error line 299718 of stdin: chromosome chrM has size 16569 but item ends at 16601\nbedGraph error line 299719 of stdin: chromosome chrM has size 16569 but item ends at 16602\nbedGraph error line 299720 of stdin: chromosome chrM has size 16569 but item ends at 16603\nbedGraph error line 299721 of stdin: chromosome chrM has size 16569 but item ends at 16605\nbedGraph error line 299722 of stdin: chromosome chrM has size 16569 but item ends at 16606\nbedGraph error line 299723 of stdin: chromosome chrM has size 16569 but item ends at 16608\nbedGraph error line 299724 of stdin: chromosome chrM has size 16569 but item ends at 16609\nbedGraph error line 299725 of stdin: chromosome chrM has size 16569 but item ends at 16610\nbedGraph error line 299726 of stdin: chromosome chrM has size 16569 but item ends at 16611\nbedGraph error line 299727 of stdin: chromosome chrM has size 16569 but item ends at 16612\nbedGraph error line 299728 of stdin: chromosome chrM has size 16569 but item ends at 16613\nbedGraph error line 299729 of stdin: chromosome chrM has size 16569 but item ends at 16614\nbedGraph error line 299730 of stdin: chromosome chrM has size 16569 but item ends at 16615\nbedGraph error line 299731 of stdin: chromosome chrM has size 16569 but item ends at 16616\nbedGraph error line 299732 of stdin: chromosome chrM has size 16569 but item ends at 16617\nbedGraph error line 299733 of stdin: chromosome chrM has size 16569 but item ends at 16618\nbedGraph error line 299734 of stdin: chromosome chrM has size 16569 but item ends at 16619\nbedGraph error line 299735 of stdin: chromosome chrM has size 16569 but item ends at 16620\nbedGraph error line 299736 of stdin: chromosome chrM has size 16569 but item ends at 16621\nbedGraph error line 299737 of stdin: chromosome chrM has size 16569 but item ends at 16622\nbedGraph error line 299738 of stdin: chromosome chrM has size 16569 but item ends at 16623\nbedGraph error line 299739 of stdin: chromosome chrM has size 16569 but item ends at 16624\nbedGraph error line 299740 of stdin: chromosome chrM has size 16569 but item ends at 16627\nbedGraph error line 299741 of stdin: chromosome chrM has size 16569 but item ends at 16629\nbedGraph error line 299742 of stdin: chromosome chrM has size 16569 but item ends at 16630\nbedGraph error line 299743 of stdin: chromosome chrM has size 16569 but item ends at 16631\nbedGraph error line 299744 of stdin: chromosome chrM has size 16569 but item ends at 16632\nbedGraph error line 299745 of stdin: chromosome chrM has size 16569 but item ends at 16633\nbedGraph error line 299746 of stdin: chromosome chrM has size 16569 but item ends at 16634\nbedGraph error line 299747 of stdin: chromosome chrM has size 16569 but item ends at 16635\nbedGraph error line 299748 of stdin: chromosome chrM has size 16569 but item ends at 16639\nbedGraph error line 299749 of stdin: chromosome chrM has size 16569 but item ends at 16641\nbedGraph error line 299750 of stdin: chromosome chrM has size 16569 but item ends at 16642\nbedGraph error line 299751 of stdin: chromosome chrM has size 16569 but item ends at 16644\nbedGraph error line 299752 of stdin: chromosome chrM has size 16569 but item ends at 16645\nbedGraph error line 299753 of stdin: chromosome chrM has size 16569 but item ends at 16649\nbedGraph error line 299754 of stdin: chromosome chrM has size 16569 but item ends at 16650\nbedGraph error line 299755 of stdin: chromosome chrM has size 16569 but item ends at 16651\nbedGraph error line 299756 of stdin: chromosome chrM has size 16569 but item ends at 16652\nbedGraph error line 299757 of stdin: chromosome chrM has size 16569 but item ends at 16653\nbedGraph error line 299758 of stdin: chromosome chrM has size 16569 but item ends at 16654\nbedGraph error line 299759 of stdin: chromosome chrM has size 16569 but item ends at 16655\nbedGraph error line 299760 of stdin: chromosome chrM has size 16569 but item ends at 16656\nbedGraph error line 299761 of stdin: chromosome chrM has size 16569 but item ends at 16657\nbedGraph error line 299762 of stdin: chromosome chrM has size 16569 but item ends at 16661\nbedGraph error line 299763 of stdin: chromosome chrM has size 16569 but item ends at 16662\nbedGraph error line 299764 of stdin: chromosome chrM has size 16569 but item ends at 16663\nbedGraph error line 299765 of stdin: chromosome chrM has size 16569 but item ends at 16664\nbedGraph error line 299766 of stdin: chromosome chrM has size 16569 but item ends at 16666\nbedGraph error line 299767 of stdin: chromosome chrM has size 16569 but item ends at 16667\n"
```
in the wig to bigWig step. That step didn't fail though, so it might be fine.

**Comment by lldelisle** [comment]
Ah yeah this is painful when it happens... (macs2 extend the reads without checking it is still into the genome...) I need to check if we get a bigwig in this case.

**Comment by lldelisle** [comment]
The bigwig is empty and I did not put a test for it. My bad.

**Comment by mvdbeek** [comment]
It's passing again

**Comment by lldelisle** [comment]
Hi,
I added a test to check the size of the bigwig and it is failing in the CI while on usegalaxy.org it is passing. I guess it is because of memory because the std error is:
```
/tmp/tmps46fzpkt/job_working_directory/000/10/tool_script.sh: line 22:     7 Done                    grep -v "^track" '/tmp/tmps46fzpkt/files/000/dataset_15.dat'
         8 Killed                  | wigToBigWig stdin '/cvmfs/data.galaxyproject.org/managed/len/ucsc/hg38.len' '/tmp/tmps46fzpkt/job_working_directory/000/10/outputs/galaxy_dataset_b50aace5-8b74-4414-a368-2fb7c073ade2.dat' -clip 2>&1
Error running wigToBigWig.
```
I guess 'Killed' means that it went out of memory.
The std output display these messages about going over the chrM size but I also get them in usegalaxy.org and the bigwig is still produced.

**Comment by mvdbeek** [comment]
> I guess 'Killed' means that it went out of memory.

It's possible, from https://genome.ucsc.edu/goldenPath/help/bigWig.html:

> Note that the wigToBigWig utility uses a substantial amount of memory: approximately 50% more memory than that of the uncompressed wiggle input file.

I think wiggle files contain base-level values for entire chromosome(s), so that could be quite large for large references, even if there are just a few mapped reads / peaks etc ... I guess ideally we'd use a test data set for a smaller reference genome, but for now we could just comment the test. I'm sure I could find some drosophila cut and run data for the tests.

**Comment by lldelisle** [comment]
I wrote another test with drosophila (very small fastqs 10MB each): https://github.com/lldelisle/iwc/pull/3

**Comment by mvdbeek** [comment]
Nice! :). Can we have both tests, and then we drop/comment the bigwig check for hg38 ?

**Comment by lldelisle** [comment]
How do we enable 2 tests?

File: workflows/epigenetics/CUTandRUN/CUTandRUN-tests.yml [hunk 1]
```diff
@@ -0,0 +1,69 @@
+- doc: Test outline for CUTandRUN
```
Comment (mvdbeek) [inline workflows/epigenetics/CUTandRUN/CUTandRUN-tests.yml:1]: This is the yaml list syntax, you can add another test like this:
```suggestion
- doc: Test outline for CUTandRUN (dm6)
   job:
     PE fastq input:
     ...
- doc: Test outline for CUTandRUN (hg38)
```

Comment (lldelisle) [reply workflows/epigenetics/CUTandRUN/CUTandRUN-tests.yml:1]: thanks

**Comment by lldelisle** [comment]
Youhou. It passes.
Ready for review/merge.

File: workflows/epigenetics/CUTandRUN/README.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
+# CUT&RUN (and CUT&TAG) Workflow
+
+## Inputs dataset
+
+- The workflow needs a single input which is a list of dataset pairs of fastqsanger.
+
+## Inputs values
+
+- adapters sequences: this depends on the library preparation. Usually CUT&RUN is Truseq and CUT&TAG is Nextera. If you don't know, use FastQC to determine if it is Truseq or Nextera
```
Comment (mvdbeek) [inline workflows/epigenetics/CUTandRUN/README.md:9]:
```suggestion
- adapter sequences: this depends on the library preparation. Usually CUT&RUN is Truseq and CUT&TAG is Nextera. If you don't know, use FastQC to determine if it is Truseq or Nextera
```

**Review by mvdbeek — APPROVED** [review]
This is awesome, my only request would be to use all lowercase for CUTandRUN, that makes it easier to type and I think it looks nicer on github

**Comment by lldelisle** [comment]
You mean use CUTandRUN instead of CUT&RUN?

**Comment by mvdbeek** [comment]
`cutandrun` or `cut-and-run` would be my vote

**Comment by mvdbeek** [comment]
https://nf-co.re/cutandrun uses cutandrun, we can still have the name of the workflow itself be `CUT&RUN`.

**Comment by lldelisle** [comment]
Youhou

**Pull request merged**